src/main/java/com/google/jenkins/plugins/credentials/oauth/AbstractGoogleRobotCredentialsDescriptor.java

```diff
@@ -55,6 +68,70 @@ public FormValidation doCheckProjectId(@QueryParameter String projectId) {
     }
   }
 
+  @CheckForNull
+  private static FormValidation checkForDuplicates(
+      String value, ModelObject context, ModelObject object) {
+    CredentialsMatcher withId = CredentialsMatchers.withId(value);
+    for (CredentialsStore store : CredentialsProvider.lookupStores(object)) {
+      if (!store.hasPermission(CredentialsProvider.VIEW)) {
+        continue;
+      }
+      ModelObject storeContext = store.getContext();
+      for (Domain domain : store.getDomains()) {
+        for (Credentials match : CredentialsMatchers.filter(store.getCredentials(domain), withId)) {
+          if (storeContext == context) {
+            return FormValidation.error("This ID is already in use");
+          } else {
+            CredentialsScope scope = match.getScope();
+            if (scope != null && !scope.isVisible(context)) {
+              // scope is not exported to child contexts
+              continue;
+            }
+            return FormValidation.warning(
+                "The ID ‘%s’ is already in use in %s",
+                value,
+                storeContext instanceof Item
+                    ? ((Item) storeContext).getFullDisplayName()
+                    : storeContext.getDisplayName());
+          }
+        }
+      }
+    }
+    return null;
+  }
+
+  public final FormValidation doCheckId(
+      @ContextInPath ModelObject context, @QueryParameter String value) {
+    if (value.isEmpty()) {
+      return FormValidation.ok();
+    }
+    if (!value.matches("[a-zA-Z0-9_.-]+")) { // anything else considered kosher?
+      return FormValidation.error("Unacceptable characters");
+    }
+    FormValidation problem = checkForDuplicates(value, context, context);
+    if (problem != null) {
+      return problem;
+    }
+    if (!(context instanceof User)) {
+      User me = User.current();
+      if (me != null) {
+        problem = checkForDuplicates(value, context, me);
+        if (problem != null) {
+          return problem;
+        }
+      }
+    }
+    if (!(context instanceof Jenkins)) {
+      // CredentialsProvider.lookupStores(User) does not return SystemCredentialsProvider.
+      Jenkins j = Jenkins.get();
+      problem = checkForDuplicates(value, context, j);
+      if (problem != null) {
+        return problem;
+      }
+    }
+    return FormValidation.ok();
+  }
+
```

Comment:
I think all this code is mostly copied from https://github.com/jenkinsci/credentials-plugin/blob/8835d63ebd8aca6548e936c7208855e4857b16f8/src/main/java/com/cloudbees/plugins/credentials/impl/BaseStandardCredentials.java#L148.

This should be inherited behaviour and should not require to write anything special on `BaseStandardCredentials` extensions.

Reply:
> I think all this code is mostly copied from https://github.com/jenkinsci/credentials-plugin/blob/8835d63ebd8aca6548e936c7208855e4857b16f8/src/main/java/com/cloudbees/plugins/credentials/impl/BaseStandardCredentials.java#L148.
> 
> This should be inherited behaviour and should not require to write anything special on `BaseStandardCredentials` extensions.

I might be wrong, but I think in our case we cannot do that, BaseStandardCredentials is using [protected static inner descriptor class](https://github.com/jenkinsci/credentials-plugin/blob/8835d63ebd8aca6548e936c7208855e4857b16f8/src/main/java/com/cloudbees/plugins/credentials/impl/BaseStandardCredentials.java#L137) and our implementations classes ([GoogleRobotPrivateKeyCredentials](https://github.com/jenkinsci/google-oauth-plugin/blob/develop/src/main/java/com/google/jenkins/plugins/credentials/oauth/GoogleRobotPrivateKeyCredentials.java#L235) and [GoogleRobotMetadataCredentials](https://github.com/jenkinsci/google-oauth-plugin/blob/develop/src/main/java/com/google/jenkins/plugins/credentials/oauth/GoogleRobotMetadataCredentials.java#L154)) are using [AbstractGoogleRobotCredentialsDescriptor](https://github.com/jenkinsci/google-oauth-plugin/blob/develop/src/main/java/com/google/jenkins/plugins/credentials/oauth/AbstractGoogleRobotCredentialsDescriptor.java#L30C23-L30C63) which is the Descriptor. We could use this [BaseStandardCredentialsDescriptor](https://github.com/jenkinsci/credentials-plugin/blob/8835d63ebd8aca6548e936c7208855e4857b16f8/src/main/java/com/cloudbees/plugins/credentials/impl/BaseStandardCredentials.java#L137) in our implementation classes as Descriptor instead of AbstractGoogleRobotCredentialsDescriptor, but we won't have any validations of AbstractGoogleRobotCredentialsDescriptor in our implementation classes.

So, we either need to use BaseStandardCredentialsDescriptor in implementation classes instead of AbstractGoogleRobotCredentialsDescriptor or somehow have common descriptor which includes both validation checks from AbstractGoogleRobotCredentialsDescriptor and BaseStandardCredentialsDescriptor.

@amuniz WDYT?

Reply:
I think if you make `AbstractGoogleRobotCredentialsDescriptor` extend from `BaseStandardCredentialsDescriptor` then that would bring all the validations "for free" without any specific code for it. This is, for example, the way [username/password credentials are implemented](https://github.com/jenkinsci/credentials-plugin/blob/8835d63ebd8aca6548e936c7208855e4857b16f8/src/main/java/com/cloudbees/plugins/credentials/impl/UsernamePasswordCredentialsImpl.java#L113).

Reply:
@amuniz IIUC the current problem is that since `AbstractGoogleRobotCredentialsDescriptor` is not an inner class to `GoogleRobotCredentials`, but its own standalone class, it can't access a `protected` class inner to `BaseStandardCredentials` (while `GoogleRobotCredentials` itself can as it extends that)?

Reply:
thank you @Riliane @amuniz, will make `AbstractGoogleRobotCredentialsDescriptor` inner class of `GoogleRobotCredentials`

Reply:
done in [19381ec](https://github.com/jenkinsci/google-oauth-plugin/pull/201/commits/19381ec976741413595a1d1316959ac5b5ea2b62)